parsers/jscript.c

```diff
@@ -1800,6 +1806,23 @@ static bool parseMethods (tokenInfo *const token, const tokenInfo *const class,
 							makeJsTag (name, JSTAG_PROPERTY, NULL, NULL);
 				}
 			}
+			else if (isType (token, TOKEN_EQUAL_SIGN)
+					 || isType (token, TOKEN_SEMICOLON)
+					 || isType (token, TOKEN_CLOSE_CURLY)
+					 || (isType (token, TOKEN_IDENTIFIER)
+						 && name->lineNumber != token->lineNumber))
```

Comment:
I didn't review this carefully, but this strikes me as a huge problem, as it means indentation affects parsing.  AFAIK, there's no JS construct that require actual line breaks. I'm not familiar with the notation this PR handles, but I'd be surprised it introduced such a fundamental language change.

Reply:
Either this would need specific handling in the prevtoken hack (speaking from memory), or record a line break has occurred so special handling can be done where needed.

In general implicit semicolons in JS is tricky to handle, because it can only be recognized on the parsing phase, not when lexing/tokenizing.  This means that to properly handle them, we should understand the *whole* JS syntax, and not the current "catchphrase-subset" we currently do (e.g. recognize specific interesting pieces and act upon that alone).  Up until now, we managed to get away with heuristics that worked in the vast majority (if not all) of the relevant cases, but maybe the field syntax breaks these, or we need to adjust it.

IIUC something like
```JS
class C0 {
    method1
    () {
    }
```
would be recognized as a `method1` field, although I'd bet JS actually thinks it's a method.

Reply:
Thank you.
After reading your review comment, I found comparing two `lineNumber` is not needed.

```
$ node --version
node --version
v14.15.4
$ cat /tmp/a.js 
cat /tmp/a.js 
class A {
    a0
    a1
}
$ cat /tmp/b.js 
cat /tmp/b.js 
class B {
    b0 b1
}

$ node /tmp/a.js 
node /tmp/a.js 
$ node /tmp/b.js 
node /tmp/b.js 
/tmp/b.js:2
    b0 b1
       ^^

SyntaxError: Unexpected identifier
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47
```

The code comparing two `lineNumber` is for rejecting `b1`.
However, the sequence of `b0 b1` is no valid JavaScript code. For such garbage input, ctags is allowed emitting garbage; JavaScript doesn't have to reject `b1` explicitly.

I will update the proposed changes.